Update Generate Prepayment Invoice form and add new details to invoice

nospame opened the pull request:
## Product Description
Updates the Generate Prepayment Invoice form with several new fields and uses these when generating the associated invoice.

![image](https://github.com/user-attachments/assets/9e4a967f-78a0-4803-809e-d52e6e866c19)


## Technical Summary
[SAAS-16745](https://dimagi.atlassian.net/browse/SAAS-16745)
Splits the knockout model that was shared between credit card prepayments and prepayment invoices in two. Adds several new fields in HTML templates for the prepayment invoice form and renames a couple of existing fields. Uses these in the `CreditsWireInvoiceView` that is accessed from the Current Subscription page, passing through to `DomainWireInvoiceFactory` to add as details on the prepayment invoice. Because of some interesting inheritance, the email fields are shared by `WireInvoiceView' on the Billing Statements page, so that view has been updated to work with the new email fields as well (though its behavior has not changed -- it will still send a separate email per email address passed in).

Recommend review by commit. Some of the bulk of here is just moving code around or mirroring B3/B5 templates.

## Safety Assurance

### Safety story
Tested locally and on staging with various valid and invalid inputs, to see that these block form submission when invalid or are populated on the invoice correctly when valid.

### Automated test coverage
Not much for this view, just for the `DomainWireInvoiceFactory`.

### QA Plan
Not planning it.


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change


[SAAS-16745]: https://dimagi.atlassian.net/browse/SAAS-16745?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nospame committed:
Move payment modal template scripts into existing partial

nospame committed:
Break prepayment invoice into initial separate form with quantity

nospame committed:
Add date fields to prepayment invoice form

nospame committed:
Remove unnecessary 'name' observable and add 'label' textfield

nospame committed:
Add input names so they get submitted with form data

nospame committed:
Comma-separate emails

nospame committed:
Add invoice_amount hidden input

nospame committed:
Use new data to create the prepayment invoice details

nospame committed:
Split CreditsWireInvoiceView validations into validate_ methods

nospame committed:
Validate prepayment invoice daterange

nospame committed:
Better display and validation for numbers

nospame committed:
Break prepayment invoice into separate form

nospame committed:
Apply changes to B5 payment_modal.html

nospame committed:
Bootstrap 5 Migration - Rebuilt diffs

nospame committed:
Update TestGetSerializableWireInvoiceItem

nospame committed:
Split prepayment invoice email field into To and CC

nospame committed:
Update WireInvoiceView to work with new fields

nospame committed:
Bootstrap 5 Migration - Rebuilt diffs

biyeun commented on corehq/apps/accounting/static/accounting/js/credits.js:
```diff
@@ -37,24 +37,38 @@ hqDefine('accounting/js/credits', [
     };
 
-    var generalCreditItem = function (paymentHandler) {
+    var ccGeneralCreditItem = function (paymentHandler) {
```
what does the `cc` stand for? I'm not super thrilled with the abbreviation here because it's not super clear what it means...

nospame replied:
That's fair, it's meant to be "credit card". I'll look at better naming along with snake_case/camelCase fixes.

nospame replied:
https://github.com/dimagi/commcare-hq/pull/36064/commits/4a6076ed8793992dfa1ffca5833eba97eeb02ebc

biyeun commented on corehq/apps/domain/templates/domain/partials/bootstrap3/payment_modal.html:
```diff
@@ -44,6 +45,42 @@
+        <div class="col-sm-10">
+          <div class="input-group">
+            <span class="input-group-addon">$</span>
+            <input type="text"
```
is this HTML new or copied? it seems to not be following the html indentation guidelines

nospame replied:
https://github.com/dimagi/commcare-hq/pull/36064/commits/7eeaac553e640af2ee6233a12e5cf8e5b93fafb2

biyeun commented on corehq/apps/domain/templates/domain/partials/bootstrap3/payment_modal.html:
```diff
@@ -72,12 +72,24 @@
+          {% trans "Quantity" %}
+        </label>
+        <div class="col-sm-10">
+          <input type="number"
```
indentation

biyeun commented on corehq/apps/accounting/static/accounting/js/credits.js:
```diff
@@ -37,24 +37,38 @@ hqDefine('accounting/js/credits', [
-        self.creditType = ko.observable("general_credit");
+        self.creditType = ko.observable("cc_general_credit");
+        self.addAmount = ko.observable(0);
+        self.addAmountValid = ko.computed(function () {
```
nit: prefer booleans to be `isAddAmountValid`

biyeun commented on corehq/apps/accounting/static/accounting/js/credits.js:
```diff
@@ -37,24 +37,38 @@ hqDefine('accounting/js/credits', [
         self.features = features;
         self.paymentHandler = paymentHandler;
-        self.general_credit = ko.observable(generalCreditItem(paymentHandler));
+        self.cc_general_credit = ko.observable(ccGeneralCreditItem(paymentHandler));
```
also if we are renaming this variable...can we camel case the variable since it's in javascript? we seem to have mixed variable naming schemes here

nospame replied:
This is a good call and will fix today. I realized I had mixed a few of these up when I put up this PR, but didn't have the brainspace to address it right then.

nospame replied:
https://github.com/dimagi/commcare-hq/pull/36064/commits/4a6076ed8793992dfa1ffca5833eba97eeb02ebc

biyeun commented on corehq/apps/accounting/static/accounting/js/payment_method_handler.js:
```diff
@@ -416,6 +416,7 @@ hqDefine('accounting/js/payment_method_handler', [
         self.features = data.features;
         self.general_credit = data.general_credit;
         self.invoice_credit = data.invoice_credit;
+        self.invoice_credit_quantity = data.invoice_credit().quantity;
```
is this quantity an observable? should this be a `ko.computed`?

nospame replied:
Yes it should (though it worked as written because I was calling `self.invoice_credit_quantity()`). Realized I probably don't even need this assignment since I'm only using it in one place where `self.invoice_credit` is already accessible.

nospame replied:
https://github.com/dimagi/commcare-hq/pull/36064/commits/4a6076ed8793992dfa1ffca5833eba97eeb02ebc

biyeun commented on corehq/apps/accounting/static/accounting/js/credits.js:
```diff
@@ -67,6 +68,8 @@ hqDefine('accounting/js/credits', [
 
     var invoiceGeneralCreditItem = function (paymentHandler) {
         var self = {};
+        self.date_start = ko.observable();
```
mixed camel case and snake case again (should be camel in JS)

biyeun commented on corehq/apps/accounting/static/accounting/js/credits.js:
```diff
@@ -128,6 +131,17 @@ hqDefine('accounting/js/credits', [
 
+    $(function () {
+        $('#paymentModal').on('shown.bs.modal', function () {
+            $("#prepay-date-start").datetimepicker({
```
can these use any existing knockout bindings? not a fan of having this format manually defined, but maybe I'm thinking of bootstrap 5 page support

nospame replied:
I think that might be B5. The places I saw `datetimepicker` used were either essentially this same thing ([here ](https://github.com/dimagi/commcare-hq/blob/de22ac74cabe41e538812da5d5d6f8967424d1b9/corehq/apps/users/static/users/js/mobile_workers.js#L545) and [here](https://github.com/dimagi/commcare-hq/blob/de22ac74cabe41e538812da5d5d6f8967424d1b9/corehq/apps/users/static/users/js/edit_commcare_user.js#L141) ) or specific to web apps.

biyeun commented on corehq/apps/accounting/static/accounting/js/credits.js:
```diff
@@ -70,7 +70,7 @@ hqDefine('accounting/js/credits', [
         self.date_start = ko.observable();
         self.date_end = ko.observable();
-        self.label = ko.observable("General Credits");
+        self.credit_label = ko.observable("General Credits");
```
mixed snake case and camel case again

nospame replied:
https://github.com/dimagi/commcare-hq/pull/36064/commits/4a6076ed8793992dfa1ffca5833eba97eeb02ebc

biyeun commented on corehq/apps/domain/templates/domain/partials/bootstrap3/payment_modal.html:
```diff
@@ -211,7 +211,7 @@
                        name="emails"
                        class="form-control"
-                       placeholder="jane@gmail.com john@gmail.com" />
+                       placeholder="jane@gmail.com, john@gmail.com" />
```
is this user facing or accounting only? if user facing can we implement a multiselect select2 widget, or is that too complicated with a bootstrap 3 page?

nospame replied:
It is user-facing, though evidently it doesn't used very often and the intent is ultimately to move it off of this page altogether and into the accounting admin interface.

Frankly, I'm not sure how to implement that widget here. I see it used with crispy forms for emails, but I didn't see anywhere it was done with html/knockout like this. I tried importing `accounting/js/widgets` making the element a `<select>` and applying the `.accounting-email-select2` class (which seemed like essentially what the crispy forms versions were doing), and it did _something_ but wasn't actually usable.

biyeun commented on corehq/apps/domain/views/accounting.py:
```diff
@@ -693,7 +693,7 @@ def dispatch(self, request, *args, **kwargs):
 
     def post(self, request, *args, **kwargs):
-        emails = request.POST.get('emails', []).split()
+        emails = request.POST.get('emails', '').replace(' ', '').split(',')
```
why did we choose to comma separate here in the first place? was there an issue with that?

nospame replied:
It was space separated before. Ops requested it be comma separated.

biyeun replied:
got it. fine with that then!

mjriley commented on corehq/apps/domain/templates/domain/partials/bootstrap3/payment_modal.html:
```diff
@@ -211,7 +211,7 @@
                        name="emails"
                        class="form-control"
-                       placeholder="jane@gmail.com john@gmail.com" />
+                       placeholder="jane@gmail.com, john@gmail.com" />
```
Does this change user behavior? How do we convey to the user that the functionality changed and they need to use a comma now?

nospame replied:
That's a good point. Would better validation help with this, something like -- if it looks like a user entered emails space-separated without a comma, we show a specific error about it?

nospame replied:
Actually, given the point [here](https://github.com/dimagi/commcare-hq/pull/36064#discussion_r2012906109) that users don't really generate their own invoices here and it's almost exclusively ops, I don't think this is that big a deal. I'd like to just have that as an invalid input by not automatically removing all whitespace.

So instead of this, which takes `email@example.com bob@dimagi.com` and reads it as `"email@example.combob@dimagi.com"` (invalid, but not what the user entered):
https://github.com/dimagi/commcare-hq/blob/1d4f5fa5912218cf7457bd465d76001ea2573c20/corehq/apps/domain/views/accounting.py#L721
We'd have this, which will still show it as invalid, but end up showing the same `"email@example.com bob@dimagi.com"` the user entered, so they can correct it:
```python
        cc_emails = [email.strip() for email in request.POST.get('email_cc', '').split(',')]
```

What do you think?

mjriley replied:
This seems reasonable to me, thanks

orangejenny commented on corehq/apps/domain/templates/domain/partials/bootstrap3/payment_modal.html:
```diff
@@ -1,5 +1,165 @@
+               data-bind="visible: !addAmountValid()">
+            <i class="fa-solid fa-triangle-exclamation"></i>
+            <span>{% blocktrans %}
+              Please enter an amount that's either $0 or  greater than $0.5
```
```suggestion
              Please enter an amount that's either $0 or  greater than $0.50
```

nospame committed:
Format payment_modal.html

nospame committed:
Variable renaming

- Clarify 'cc' and 'cardPayment' and 'invoice' as 'prepaymentInvoice'
- Use camelCase for JS variables
- Don't assign self.invoice_credit_quantity"

nospame committed:
Lint python

nospame committed:
Two decimal points for dollar amounts

nospame committed:
Remove whitespace from cc_emails after splitting

nospame committed:
Update one use of control-label -> form-label

nospame committed:
Bootstrap 5 Migration - Rebuilt diffs

biyeun reviewed (approved):
this looks good. I ended up reviewing files changed as a whole to get a big picture idea of what is going on. trusted you tested locally and on staging well enough for accounting use cases
